File: phantom-windows-app/Services/CompanionRelayClient.cs
using System;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using SecureOverlay.Infrastructure.Hosted.Contracts;

namespace SecureOverlay.Services
{
    /// <summary>
    /// WebSocket relay client for Companion Mode. Connects to the hosted relay with a
    /// short-lived ticket, sends/receives JSON envelopes, and reconnects with a capped
    /// backoff. A fresh ticket is fetched from <see cref="_ticketProvider"/> on every
    /// connect attempt (spec §9) so reconnects after a drop do not fail on a consumed
    /// ticket. All inbound frames are surfaced via <see cref="FrameReceived"/>; the
    /// caller (CompanionCommandHost) decides how to react.
    /// </summary>
    public sealed class CompanionRelayClient : IAsyncDisposable
    {
        private const string SubProtocol = "phantom.companion.v1";
        private static readonly TimeSpan ReceiveTimeout = TimeSpan.FromSeconds(90);
        private static readonly TimeSpan PingInterval = TimeSpan.FromSeconds(15);
        private const int MaxAssemblyBytes = 128 * 1024;
        private static readonly JsonSerializerOptions JsonOptions = new()
        {
            PropertyNameCaseInsensitive = true
        };
        private static readonly TimeSpan[] Backoff =
        {
            TimeSpan.FromSeconds(1),
            TimeSpan.FromSeconds(2),
            TimeSpan.FromSeconds(4),
            TimeSpan.FromSeconds(8),
            TimeSpan.FromSeconds(15)
        };

        private readonly Func<CancellationToken, Task<CompanionRelayTicketDto>> _ticketProvider;
        private ClientWebSocket? _socket;
        private CancellationTokenSource _cts = new();
        private Task? _loopTask;
        private int _backoffIndex;

        public string PairingId { get; }
        public string Role { get; }
        public bool IsRunning => _loopTask != null && !_loopTask.IsCompleted;

        public event Action<CompanionRelayFrame>? FrameReceived;
        public event Action<CompanionRelayState>? StateChanged;

        public CompanionRelayClient(Func<CancellationToken, Task<CompanionRelayTicketDto>> ticketProvider, string pairingId, string role)
        {
            _ticketProvider = ticketProvider;
            PairingId = pairingId;
            Role = role;
        }

        public void Start()
        {
            if (IsRunning) return;
            _cts = new CancellationTokenSource();
            _loopTask = Task.Run(() => RunAsync(_cts.Token));
        }

        public async Task StopAsync()
        {
            _cts.Cancel();
            if (_socket != null && _socket.State == WebSocketState.Open)
            {
                try { await _socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "stopping", CancellationToken.None); } catch { }
            }
            if (_loopTask != null)
            {
                try { await _loopTask; } catch { }
            }
            _loopTask = null;
        }

        public async ValueTask DisposeAsync()
        {
            await StopAsync();
            _cts.Dispose();
        }

        public async Task SendAsync(CompanionRelayFrame frame, CancellationToken cancellationToken = default)
        {
            var socket = _socket;
            if (socket == null || socket.State != WebSocketState.Open) return;
            var payload = JsonSerializer.SerializeToUtf8Bytes(frame);
            await socket.SendAsync(payload, WebSocketMessageType.Text, endOfMessage: true, cancellationToken);
        }

        public async Task SendRawAsync(string json, CancellationToken cancellationToken = default)
        {
            var socket = _socket;
            if (socket == null || socket.State != WebSocketState.Open) return;
            var payload = Encoding.UTF8.GetBytes(json);
            await socket.SendAsync(payload, WebSocketMessageType.Text, endOfMessage: true, cancellationToken);
        }

        private async Task RunAsync(CancellationToken cancellationToken)
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                EmitState(CompanionRelayState.Connecting);
                try
                {
                    // Fetch a fresh ticket on every connect (spec §9). A ticket is
                    // single-use and consumed by the backend on upgrade, so reusing a
                    // stored ticket would permanently break reconnects.
                    CompanionRelayTicketDto ticket;
                    try
                    {
                        ticket = await _ticketProvider(cancellationToken);
                    }
                    catch (Exception ex)
                    {
                        Log.WriteLine($"CompanionRelayClient ticket fetch failed: {ex.Message}");
                        throw;
                    }

                    using var socket = new ClientWebSocket();
                    socket.Options.AddSubProtocol(SubProtocol);
                    var uriBuilder = new UriBuilder(ticket.RelayUrl)
                    {
                        Query = $"ticket={Uri.EscapeDataString(ticket.Ticket)}"
                    };
                    await socket.ConnectAsync(uriBuilder.Uri, cancellationToken);
                    _socket = socket;
                    _backoffIndex = 0;
                    EmitState(CompanionRelayState.Connected);
                    await ReceiveLoopAsync(socket, cancellationToken);
                }
                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
                {
                    break;
                }
                catch (Exception ex)
                {
                    Log.WriteLine($"CompanionRelayClient connection error: {ex.Message}");
                    if (IsTerminalPairingFailure(ex.Message))
                    {
                        FrameReceived?.Invoke(new CompanionRelayFrame
                        {
                            V = 1,
                            Type = "relay.error",
                            Code = "pairing_revoked",
                            Message = ex.Message
                        });
                        break;
                    }
                }
                finally
                {
                    _socket = null;
                    EmitState(CompanionRelayState.Disconnected);
                }

                if (cancellationToken.IsCancellationRequested) break;
                var delay = Backoff[Math.Min(_backoffIndex, Backoff.Length - 1)];
                _backoffIndex = Math.Min(_backoffIndex + 1, Backoff.Length - 1);
                try { await Task.Delay(delay, cancellationToken); } catch { break; }
            }
        }

        private async Task ReceiveLoopAsync(ClientWebSocket socket, CancellationToken cancellationToken)
        {
            var buffer = new byte[MaxAssemblyBytes];
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

            // Outbound ping every 20s (spec §5: client may ping; server drops after 45s silence).
            var ping = Task.Run(async () =>
            {
                try
                {
                    while (!cts.IsCancellationRequested && socket.State == WebSocketState.Open)
                    {
                        await Task.Delay(PingInterval, cts.Token);
                        if (socket.State == WebSocketState.Open)
                        {
                            var pingPayload = JsonSerializer.SerializeToUtf8Bytes(new
                            {
                                v = 1, type = "relay.ping", ts = DateTime.UtcNow
                            });
                            await socket.SendAsync(pingPayload, WebSocketMessageType.Text, endOfMessage: true, cts.Token);
                        }
                    }
                }
                catch { }
            }, cts.Token);

            try
            {
                while (socket.State == WebSocketState.Open && !cancellationToken.IsCancellationRequested)
                {
                    // 45s receive timeout — drop silent sockets.
                    cts.CancelAfter(ReceiveTimeout);
                    WebSocketReceiveResult result;
                    using var ms = new System.IO.MemoryStream();
                    var tooLarge = false;
                    do
                    {
                        result = await socket.ReceiveAsync(buffer, cts.Token);
                        if (result.MessageType == WebSocketMessageType.Close)
                        {
                            await socket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "client closing", CancellationToken.None);
                            return;
                        }
                        ms.Write(buffer, 0, result.Count);
                        if (ms.Length > MaxAssemblyBytes)
                        {
                            tooLarge = true;
                            while (!result.EndOfMessage)
                            {
                                result = await socket.ReceiveAsync(buffer, cts.Token);
                            }
                            break;
                        }
                    }
                    while (!result.EndOfMessage);

                    if (tooLarge)
                    {
                        Log.WriteLine("CompanionRelayClient: inbound frame exceeded size limit; ignored.");
                        continue;
                    }

                    if (ms.Length == 0) continue;

                    CompanionRelayFrame? frame;
                    try
                    {
                        frame = JsonSerializer.Deserialize<CompanionRelayFrame>(ms.ToArray(), JsonOptions);
                    }
                    catch (Exception ex)
                    {
                        Log.WriteLine($"CompanionRelayClient failed to parse frame: {ex.Message}");
                        continue;
                    }

                    if (frame != null)
                    {
                        if (string.Equals(frame.Type, "relay.ping", StringComparison.Ordinal))
                        {
                            var pongPayload = JsonSerializer.SerializeToUtf8Bytes(new
                            {
                                v = 1, type = "relay.pong", ts = DateTime.UtcNow,
                                pairingId = PairingId, role = Role
                            });
                            try
                            {
                                await socket.SendAsync(pongPayload, WebSocketMessageType.Text, endOfMessage: true, cancellationToken);
                            }
                            catch { }
                            continue;
                        }
                        FrameReceived?.Invoke(frame);
                    }
                }
            }
            finally
            {
                cts.Cancel();
                try { await ping; } catch { }
            }
        }

        private void EmitState(CompanionRelayState state) => StateChanged?.Invoke(state);

        private static bool IsTerminalPairingFailure(string message)
        {
            if (string.IsNullOrWhiteSpace(message)) return false;
            return message.Contains("revoked", StringComparison.OrdinalIgnoreCase)
                || message.Contains("not authorized", StringComparison.OrdinalIgnoreCase)
                || message.Contains("Pairing not found", StringComparison.OrdinalIgnoreCase)
                || message.Contains("no longer available", StringComparison.OrdinalIgnoreCase);
        }
    }

    public enum CompanionRelayState { Connecting, Connected, Disconnected }

    /// <summary>Outcome of a companion-driven chat turn, used to emit the correct
    /// chat.* frame from the SendMessage finally block (C5).</summary>
    public enum CompanionTurnOutcome { Pending, Succeeded, Cancelled, Failed }

    public sealed class CompanionRelayFrame
    {
        public int V { get; set; } = 1;
        public string? Id { get; set; }
        public string? Type { get; set; }
        public DateTime? Ts { get; set; }
        public string? PairingId { get; set; }
        public string? Role { get; set; }
        // Top-level fields used by relay.error (code/message are sent at the envelope top
        // level, not under body). Bound via case-insensitive deserialization.
        public string? Code { get; set; }
        public string? Message { get; set; }
        public JsonElement? Body { get; set; }

        // Reads a string field from the envelope BODY (where phone→desktop commands live).
        public string? ReadString(string key)
            => Body.HasValue && Body.Value.ValueKind == JsonValueKind.Object
                && Body.Value.TryGetProperty(key, out var p) && p.ValueKind == JsonValueKind.String
                ? p.GetString() : null;

        public bool? ReadBool(string key)
            => Body.HasValue && Body.Value.ValueKind == JsonValueKind.Object
                && Body.Value.TryGetProperty(key, out var p) && p.ValueKind == JsonValueKind.True
                ? true
                : (Body.HasValue && Body.Value.ValueKind == JsonValueKind.Object
                    && Body.Value.TryGetProperty(key, out var q) && q.ValueKind == JsonValueKind.False ? false : null);

        public int? ReadInt(string key)
        {
            if (!Body.HasValue || Body.Value.ValueKind != JsonValueKind.Object) return null;
            if (!Body.Value.TryGetProperty(key, out var p)) return null;
            return p.ValueKind switch
            {
                JsonValueKind.Number when p.TryGetInt32(out var n) => n,
                JsonValueKind.String when int.TryParse(p.GetString(), out var parsed) => parsed,
                _ => null
            };
        }
    }
}

File: phantom-mac-app/Sources/Phantom/CompanionRelayClient.swift
import Foundation

/// WebSocket relay client for Companion Mode on macOS. Connects to the hosted relay with a
/// short-lived ticket using URLSessionWebSocketTask, sends/receives JSON envelopes, and
/// reconnects with a capped backoff. A fresh ticket is fetched from `ticketProvider` on
/// every connect attempt (spec §9) so reconnects after a drop do not fail on a consumed
/// ticket. Inbound frames are surfaced via `frameHandler`.
@MainActor
final class CompanionRelayClient {
    private static let subprotocol = "phantom.companion.v1"
    private static let receiveTimeout: TimeInterval = 90
    private static let pingInterval: TimeInterval = 15
    private static let backoff: [TimeInterval] = [1, 2, 4, 8, 15]

    private let ticketProvider: () async throws -> CompanionRelayTicket
    let pairingId: String
    private let role: String
    // Reuse a single URLSession across reconnects to avoid leaking sessions (M15).
    private let session = URLSession(configuration: .default)
    private var task: URLSessionWebSocketTask?
    private var pingTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?
    private var watchdogTask: Task<Void, Never>?
    private var backoffIndex = 0
    private var stopped = false
    private var lastReceivedAt = Date()

    private(set) var state: CompanionRelayState = .disconnected

    var frameHandler: ((CompanionRelayFrame) -> Void)?
    var stateHandler: ((CompanionRelayState) -> Void)?

    init(ticketProvider: @escaping () async throws -> CompanionRelayTicket, pairingId: String, role: String) {
        self.ticketProvider = ticketProvider
        self.pairingId = pairingId
        self.role = role
    }

    func start() {
        stopped = false
        connect()
    }

    func stop() {
        stopped = true
        pingTask?.cancel()
        receiveTask?.cancel()
        watchdogTask?.cancel()
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        session.invalidateAndCancel()
        updateState(.disconnected)
    }

    func send(frame: CompanionRelayFrame) {
        guard let task, task.state == .running else { return }
        do {
            let data = try JSONEncoder().encode(frame)
            task.send(.data(data)) { error in
                if let error {
                    print("[companion] relay send error: \(error.localizedDescription)")
                }
            }
        } catch {
            print("[companion] relay encode error: \(error.localizedDescription)")
        }
    }

    func sendRaw(_ json: String) {
        guard let task, task.state == .running, let data = json.data(using: .utf8) else { return }
        task.send(.data(data)) { _ in }
    }

    private func connect() {
        guard !stopped else { return }
        updateState(.connecting)
        Task { [weak self] in
            guard let self else { return }
            let ticket: CompanionRelayTicket
            do {
                ticket = try await self.ticketProvider()
            } catch {
                print("[companion] relay ticket fetch failed: \(error.localizedDescription)")
                if Self.isTerminalPairingFailure(error.localizedDescription) {
                    self.stopped = true
                    var errorFrame = CompanionRelayFrame()
                    errorFrame.type = "relay.error"
                    errorFrame.code = "pairing_revoked"
                    errorFrame.message = error.localizedDescription
                    self.frameHandler?(errorFrame)
                    self.updateState(.disconnected)
                    return
                }
                if !self.stopped { self.scheduleReconnect() }
                return
            }
            guard !self.stopped else { return }
            self.openSocket(with: ticket)
        }
    }

    private func openSocket(with ticket: CompanionRelayTicket) {
        guard let url = URL(string: ticket.relayUrl) else {
            print("[companion] bad relay url: \(ticket.relayUrl)")
            updateState(.disconnected)
            scheduleReconnect()
            return
        }
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        var queryItems = components?.queryItems ?? []
        queryItems.append(URLQueryItem(name: "ticket", value: ticket.ticket))
        components?.queryItems = queryItems
        guard let ticketedUrl = components?.url else {
            print("[companion] bad relay url")
            updateState(.disconnected)
            scheduleReconnect()
            return
        }
        let wsTask = session.webSocketTask(with: ticketedUrl, protocols: [Self.subprotocol])
        task = wsTask
        wsTask.resume()
        backoffIndex = 0
        lastReceivedAt = Date()
        // Stay `.connecting` until the first frame confirms the handshake (M14).
        startPing()
        startReceive()
        startWatchdog()
    }

    private func startPing() {
        pingTask?.cancel()
        pingTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled, !self.stopped {
                try? await Task.sleep(nanoseconds: UInt64(Self.pingInterval * 1_000_000_000))
                guard !self.stopped, self.task != nil else { break }
                let ts = ISO8601DateFormatter().string(from: Date())
                self.sendRaw("{\"v\":1,\"type\":\"relay.ping\",\"ts\":\"\(ts)\",\"pairingId\":\"\(self.pairingId)\",\"role\":\"\(self.role)\"}")
            }
        }
    }

    private func startReceive() {
        receiveTask?.cancel()
        receiveTask = Task { [weak self] in
            guard let self else { return }
            var handshakeConfirmed = false
            while !Task.isCancelled, !self.stopped {
                guard let task = self.task, task.state == .running else { break }
                do {
                    let message = try await task.receive()
                    self.lastReceivedAt = Date()
                    if !handshakeConfirmed {
                        handshakeConfirmed = true
                        self.updateState(.connected)
                    }
                    switch message {
                    case .data(let data):
                        self.handle(data: data)
                    case .string(let text):
                        if let data = text.data(using: .utf8) {
                            self.handle(data: data)
                        }
                    @unknown default:
                        break
                    }
                } catch {
                    break
                }
            }
            if !self.stopped {
                self.updateState(.disconnected)
                self.scheduleReconnect()
            }
        }
    }

    // Watchdog: if no frame (or ping) is received within 45s, the socket is dead —
    // cancel it so the receive loop exits and we reconnect (M13).
    private func startWatchdog() {
        watchdogTask?.cancel()
        watchdogTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled, !self.stopped {
                try? await Task.sleep(nanoseconds: 5_000_000_000)
                if self.stopped { break }
                if Date().timeIntervalSince(self.lastReceivedAt) >= Self.receiveTimeout {
                    if let task = self.task, task.state == .running {
                        task.cancel(with: .abnormalClosure, reason: "receive timeout".data(using: .utf8))
                    }
                    break
                }
            }
        }
    }

    private func handle(data: Data) {
        do {
            let frame = try JSONDecoder().decode(CompanionRelayFrame.self, from: data)
            frameHandler?(frame)
        } catch {
            print("[companion] relay decode error: \(error.localizedDescription)")
        }
    }

    private func scheduleReconnect() {
        guard !stopped else { return }
        let delay = Self.backoff[min(backoffIndex, Self.backoff.count - 1)]
        backoffIndex = min(backoffIndex + 1, Self.backoff.count - 1)
        Task { [weak self] in
            guard let self else { return }
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard !self.stopped else { return }
            self.connect()
        }
    }

    private func updateState(_ newState: CompanionRelayState) {
        state = newState
        stateHandler?(newState)
    }

    private static func isTerminalPairingFailure(_ message: String) -> Bool {
        let text = message.lowercased()
        return text.contains("revoked")
            || text.contains("not authorized")
            || text.contains("pairing not found")
            || text.contains("no longer available")
    }
}

enum CompanionRelayState { case connecting, connected, disconnected }

struct CompanionRelayFrame: Codable {
    var v: Int = 1
    var id: String?
    var type: String?
    // `ts` is kept as a String so an unexpected date format from the relay never breaks
    // frame decoding (the client does not interpret `ts`). (C2)
    var ts: String?
    var pairingId: String?
    var role: String?
    // Top-level fields used by relay.error (code/message are sent at the envelope top
    // level, not under body). (M18)
    var code: String?
    var message: String?
    var body: AnyCodable?

    func string(_ key: String) -> String? {
        body?.value(key) as? String
    }

    func bool(_ key: String) -> Bool? {
        body?.value(key) as? Bool
    }
}

/// Lightweight AnyCodable wrapper so the relay frame body can carry arbitrary JSON without
/// forcing a fixed schema on the desktop client.
struct AnyCodable: Codable {
    private let valueBox: Any

    init(_ value: Any) { valueBox = value }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let dict = try? container.decode([String: AnyCodable].self) {
            // Lowercase dict keys so body field reads are case-insensitive (L3 hardening).
            // The phone sends camelCase today, but this protects against any casing drift.
            valueBox = Dictionary(uniqueKeysWithValues: dict.map { ($0.key.lowercased(), $0.value) })
        } else if let string = try? container.decode(String.self) {
            valueBox = string
        } else if let number = try? container.decode(Double.self) {
            valueBox = number
        } else if let bool = try? container.decode(Bool.self) {
            valueBox = bool
        } else {
            valueBox = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let dict = valueBox as? [String: AnyCodable] {
            try container.encode(dict)
        } else if let string = valueBox as? String {
            try container.encode(string)
        } else if let number = valueBox as? Double {
            try container.encode(number)
        } else if let bool = valueBox as? Bool {
            try container.encode(bool)
        } else {
            try container.encodeNil()
        }
    }

    func value(_ key: String) -> Any? {
        if let dict = valueBox as? [String: AnyCodable], let entry = dict[key.lowercased()] {
            return entry.valueBox
        }
        return nil
    }
}
